File: frontend/src/api/github.ts
import { request } from "./core";
import type { GithubReleasesResponse } from "../types/generated/GithubReleasesResponse";
import type { GithubSettingsResponse } from "../types/generated/GithubSettingsResponse";
import type { SaveGithubSettingsRequest } from "../types/generated/SaveGithubSettingsRequest";

/**
 * Fetch GitHub release tags for a repository.
 *
 * @param repo - GitHub repository in "owner/repo" format
 * @returns List of release tags sorted by published date (oldest first)
 */
export async function fetchGithubReleases(
  repo: string,
): Promise<GithubReleasesResponse> {
  return request<GithubReleasesResponse>(
    "GET",
    `/github/releases?repo=${encodeURIComponent(repo)}`,
  );
}

/**
 * Get GitHub settings (admin only).
 * Returns whether a token is configured without revealing the actual token.
 *
 * @returns GitHub settings response with has_token flag
 */
export async function getGithubSettings(): Promise<GithubSettingsResponse> {
  return request<GithubSettingsResponse>("GET", "/admin/settings/github");
}

/**
 * Save GitHub settings (admin only).
 *
 * @param settings - The settings to save (api_token)
 */
export async function saveGithubSettings(
  settings: SaveGithubSettingsRequest,
): Promise<void> {
  await request<void>("PUT", "/admin/settings/github", settings);
}

File: backend/src/api/github.rs
//! API endpoints for GitHub integration.

use std::sync::Arc;

use axum::extract::{Query, State};
use axum::http::StatusCode;
use axum::Json;

use crate::auth::AuthUser;
use crate::error::AppError;
use crate::integrations::github;
use crate::types::system::{
    GithubReleasesQuery, GithubReleasesResponse, GithubSettings, GithubSettingsResponse,
    SaveGithubSettingsRequest,
};
use crate::types::Role;
use crate::AppState;

/// GET /api/github/releases
///
/// Fetch release tags for a GitHub repository.
/// Uses cached results when available to avoid hitting rate limits.
pub async fn get_releases(
    State(state): State<Arc<AppState>>,
    _user: AuthUser,
    Query(query): Query<GithubReleasesQuery>,
) -> Result<Json<GithubReleasesResponse>, AppError> {
    // Get GitHub settings to check for token
    let github_settings = state.db.get_github_settings().await?;
    let token = github_settings.and_then(|s| s.api_token);

    // Fetch releases from GitHub
    let releases =
        github::fetch_release_tags(&state.http_client, &query.repo, token.as_deref()).await?;

    Ok(Json(GithubReleasesResponse {
        releases,
        cached: false,
    }))
}

/// GET /api/admin/settings/github
///
/// Get GitHub settings (admin only).
/// Returns whether a token is configured without revealing the actual token.
pub async fn get_github_settings(
    State(state): State<Arc<AppState>>,
    user: AuthUser,
) -> Result<Json<GithubSettingsResponse>, AppError> {
    if user.role != Role::Admin {
        return Err(AppError::Forbidden("Admin access required".to_string()));
    }

    let settings = state.db.get_github_settings().await?;
    let has_token = settings.and_then(|s| s.api_token).is_some();

    Ok(Json(GithubSettingsResponse { has_token }))
}

/// PUT /api/admin/settings/github
///
/// Save GitHub settings (admin only).
pub async fn save_github_settings(
    State(state): State<Arc<AppState>>,
    user: AuthUser,
    Json(req): Json<SaveGithubSettingsRequest>,
) -> Result<StatusCode, AppError> {
    if user.role != Role::Admin {
        return Err(AppError::Forbidden("Admin access required".to_string()));
    }

    // If token is empty or None, delete the settings
    if req.api_token.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        state.db.delete_github_settings().await?;
        return Ok(StatusCode::NO_CONTENT);
    }

    // Save the settings
    let settings = GithubSettings {
        api_token: req.api_token,
    };
    state.db.save_github_settings(&settings).await?;

    Ok(StatusCode::NO_CONTENT)
}
